optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java

```diff
@@ -177,4 +200,59 @@ public Duration getSolvingDuration() {
         }
         return Duration.ofMillis(endingSystemTimeMillis - startingSystemTimeMillis);
     }
+
+    /**
+     * A listener that unlocks the solverStatusModifyingLock when Solving has started.
+     *
+     * It to prevent the following scenario caused by unlocking before Solving started:
+     *
+     * Thread 1:
+     * solverStatusModifyingLock.unlock()
+     * >solver.solve(...) // executes second
+     *
+     * Thread 2:
+     * case SOLVING_ACTIVE:
+     * >solver.terminateEarly(); // executes first
+     *
+     * The solver.solve() call resets the terminateEarly flag, and thus the solver will not be terminated
+     * by the call, which means terminatedLatch will not be decremented, causing Thread 2 to wait forever
+     * (at least until another Thread calls terminateEarly again).
+     *
+     * To prevent Thread 2 from potentially waiting forever, we only unlock the lock after the
+     * solvingStarted phase lifecycle event is fired, meaning the terminateEarly flag will not be
+     * reset and thus the solver will actually terminate.
+     */
+    private final class UnlockLockPhaseLifecycleListener implements PhaseLifecycleListener<Solution_> {
```

Comment:
We have the `PhaseLifecycleListenerAdapter`. You may want to extend that and only override `solvingStarted()` and skip the "Do nothing" overrides.